lib/AST/GenericSignature.cpp

```diff
@@ -994,17 +776,74 @@ GenericSignatureImpl::getConformanceAccessPath(Type type,
   auto conforms = equivClass->conformsTo.find(protocol);
   assert(conforms != equivClass->conformsTo.end());
 
-  // Canonicalize the root type.
-  auto source = getBestRequirementSource(builder, conforms->second);
-  Type rootType = source->getRootType()->getCanonicalType(this);
+  SmallVector<std::tuple<Type, Type, ProtocolDecl *>, 2> derivedConstraints;
+  for (auto constraint : conforms->second) {
```

Comment:
This is quite straightforward and clear. Why the two-pass algorithm, gathering derived constraints and then walking them again later, when one pass would suffice? Is this staging for an algorithmic change?

Reply:
Mostly because I haven't convinced myself that the current 'best requirement' heuristic here is adequate. I was thinking it might be necessary to canonicalize the parent types and pick the best canonical one, for example. However, I see that even that can be done in a single pass without building `derivedConstraints`. I'll revisit this part.

Reply:
Yeah, I was imagining I would need to sort, but clearly that's not necessary, we just need the minimum element. I'll also add comments explaining the recursion and base case later.